source/common/stats/stat_merger.cc

```diff
@@ -0,0 +1,84 @@
+#include "common/stats/stat_merger.h"
+
+namespace Envoy {
+namespace Stats {
+
+StatMerger::StatMerger(Stats::Store& target_store) : target_store_(target_store) {}
+
+void StatMerger::mergeStats(const Protobuf::Map<std::string, uint64_t>& counters,
+                            const Protobuf::Map<std::string, uint64_t>& gauges) {
+  for (const auto& counter : counters) {
+    uint64_t new_parent_value = counter.second;
+    auto found_value = parent_counter_values_.find(counter.first);
+    uint64_t old_parent_value =
+        found_value == parent_counter_values_.end() ? 0 : found_value->second;
+    parent_counter_values_[counter.first] = new_parent_value;
+    target_store_.counter(counter.first).add(new_parent_value - old_parent_value);
+  }
+
+  // Gauge name *substrings*, and special logic to use for combining those gauges' values.
+  static const std::vector<std::pair<std::string, CombineLogic>> combine_logic_exceptions{
+      {".version", CombineLogic::NoImport},
+      {".connected_state", CombineLogic::BooleanOr},
+      {"server.live", CombineLogic::NoImport},
+      {"runtime.admin_overrides_active", CombineLogic::NoImport},
+      {"runtime.num_keys", CombineLogic::NoImport},
+      {"cluster_manager.active_clusters", CombineLogic::OnlyImportWhenUnused},
+      {"cluster_manager.warming_clusters", CombineLogic::OnlyImportWhenUnused},
+      {".membership_total", CombineLogic::OnlyImportWhenUnused},
+      {".membership_healthy", CombineLogic::OnlyImportWhenUnused},
+      {".membership_degraded", CombineLogic::OnlyImportWhenUnused},
+      {".max_host_weight", CombineLogic::OnlyImportWhenUnused},
+      {".total_principals", CombineLogic::OnlyImportWhenUnused},
+      {"listener_manager.total_listeners_draining", CombineLogic::NoImport},
+      {"listener_manager.total_listeners_warming", CombineLogic::OnlyImportWhenUnused},
+      {"listener_manager.total_listeners_active", CombineLogic::OnlyImportWhenUnused},
+      {"pressure", CombineLogic::OnlyImportWhenUnused},
+      {"server.concurrency", CombineLogic::OnlyImportWhenUnused},
+      {"server.hot_restart_epoch", CombineLogic::NoImport},
+  };
+  std::cerr << "MERGING GAUGES:" << std::endl;
```

Comment:
del or convert to envoy logging, same elsewhere if I missed anything else.

Reply:
Done.